pkg/kubelet/kubeletconfig/util/codec/codec.go

```diff
@@ -72,16 +77,39 @@ func NewYAMLEncoder(groupName string) (runtime.Encoder, error) {
 		return nil, fmt.Errorf("no enabled versions for group %q", groupName)
 	}
 
-	// the "best" version supposedly comes first in the list returned from legacyscheme.Registry.EnabledVersionsForGroup
+	// the "best" version supposedly comes first in the list returned from legacyscheme.Registry.EnabledVersionsForGroup.
 	return legacyscheme.Codecs.EncoderForVersion(info.Serializer, versions[0]), nil
 }
 
-// DecodeKubeletConfiguration decodes a serialized KubeletConfiguration to the internal type
+// DecodeKubeletConfiguration decodes a serialized KubeletConfiguration to the internal type.
 func DecodeKubeletConfiguration(kubeletCodecs *serializer.CodecFactory, data []byte) (*kubeletconfig.KubeletConfiguration, error) {
-	// the UniversalDecoder runs defaulting and returns the internal type by default
+	var (
+		obj runtime.Object
+		gvk *schema.GroupVersionKind
+	)
+
+	// The UniversalDecoder runs defaulting and returns the internal type by default.
 	obj, gvk, err := kubeletCodecs.UniversalDecoder().Decode(data, nil, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to decode, error: %v", err)
+		// Try strict decoding first. If that fails decode with a lenient
+		// decoder, which has only v1beta1 registered, and log a warning.
+		// The lenient path is to be dropped when v1beta1 is deprecated.
+		if !runtime.IsStrictDecodingError(err) {
+			return nil, errors.Wrap(err, "failed to decode")
+		}
+		klog.Warningf("using lenient decoding as strict decoding failed: %v", err)
+		lenientScheme := runtime.NewScheme()
```

Comment:
prepare the lenient scheme outside this method (as a private package variable is fine) and reuse it